lib/spack/spack/environment/environment.py

```diff
@@ -1743,47 +1744,34 @@ def _read_lockfile(self, file_or_json):
 
     def _read_lockfile_dict(self, d):
         """Read a lockfile dictionary into this environment."""
+        self.specs_by_hash = {}
+
```

Comment:
I think it's clearer, whenever possible, to initialize attributes in `__init__`. I usually find it "surprising" when, after calling an object method, a new public attribute pops into existence.

Reply:
That makes sense, I usually go with that approach too.  For the record, I didn't introduce that in this PR, just moved the initialization to the top of the method from somewhere in the middle.  In general, this class seems to violate that principle quite a lot, e.g. the `clear` method resets a bunch of container members, almost none of which are mentioned in `__init__`.  I could move this up to `__init__`, I'm just worried it may quietly break something.

Reply:
This is already initialized in `__init__` via `clear()`